manta/src/main/java/ch/cyberduck/core/manta/MantaExceptionMappingService.java

```diff
@@ -0,0 +1,115 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+import ch.cyberduck.core.AbstractExceptionMappingService;
+import ch.cyberduck.core.DefaultIOExceptionMappingService;
+import ch.cyberduck.core.LocaleFactory;
+import ch.cyberduck.core.Path;
+import ch.cyberduck.core.StringAppender;
+import ch.cyberduck.core.exception.AccessDeniedException;
+import ch.cyberduck.core.exception.BackgroundException;
+import ch.cyberduck.core.exception.InteroperabilityException;
+import ch.cyberduck.core.exception.LoginFailureException;
+import ch.cyberduck.core.exception.NotfoundException;
+import ch.cyberduck.core.http.HttpResponseExceptionMappingService;
+
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.http.client.HttpResponseException;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.security.KeyException;
+
+import com.joyent.manta.exception.MantaClientException;
+import com.joyent.manta.exception.MantaClientHttpResponseException;
+import com.joyent.manta.exception.MantaException;
+import com.joyent.manta.exception.MantaIOException;
+
+public class MantaExceptionMappingService extends AbstractExceptionMappingService<Exception> {
+
+    private static final Logger log = Logger.getLogger(MantaExceptionMappingService.class);
+    private final MantaSession session;
+
+    public MantaExceptionMappingService(final MantaSession session) {
+        this.session = session;
+    }
+
+    public MantaExceptionMappingService() {
+        this.session = null;
+    }
+
+    @Override
+    public BackgroundException map(final Exception failure) {
+
+        // StringWriter sw = new StringWriter();
+        // PrintWriter pw = new PrintWriter(sw);
+        // failure.printStackTrace(pw);
+        // log.error(sw.toString());
+
+        /**
+         * TODO: more fine-grained mapping
+         */
+
+        if(failure instanceof KeyException) {
+            return new LoginFailureException("Login failed", failure);
+        }
+        if(failure instanceof MantaClientHttpResponseException) {
+            final MantaClientHttpResponseException me = (MantaClientHttpResponseException) failure;
+            return mapHttpException(me, ObjectUtils.firstNonNull(me.getContent(), ""));
+        }
+
+        if(failure instanceof MantaIOException) {
+            return new DefaultIOExceptionMappingService().map((IOException) ExceptionUtils.getRootCause(failure));
+        }
+        return new InteroperabilityException(failure.getMessage(), failure);
+    }
+
+    public BackgroundException mapHttpException(final MantaClientHttpResponseException httpFailure, final String path) {
```

Comment:
You can possibly resuse `HttpResponseExceptionMappingService` with `return new HttpResponseExceptionMappingService().map(new HttpResponseException(httpFailure.getStatusCode(), buffer.toString()))`.